### PR TITLE
Composite data type find handling

### PR DESCRIPTION
**Pull request by ribaraka**

### CODE REVIEW

File: integration/shareddata/composites.go [hunk 1]
```diff
@@ -21,11 +21,13 @@ import "go.mongodb.org/mongo-driver/bson"
 // This shared data set is not frozen yet, but please add to it only if it is really shared.
 var Composites = &Values[string]{
 	data: map[string]any{
-		"document":       bson.D{{"foo", int32(42)}},
-		"document-empty": bson.D{},
+		"document":            bson.D{{"foo", int32(42)}},
+		"document-resembling": bson.D{{"foo", int32(42)}, {"42", "foo"}, {"array", bson.A{"foo", int32(42)}}},
```
Comment (seeforschauer) [inline integration/shareddata/composites.go:25]: what means resembling? what to resemble? can we please specify more details?

**Review by seeforschauer — CHANGES_REQUESTED** [review]
check dirty and some tests you've added are not ok, can you please fix it?

File: integration/shareddata/composites.go (hunk 1 above)
Comment (ribaraka) [reply integration/shareddata/composites.go:25]: "document-resembling" is resembling the existed array and document above. I've created this for checking for equivalent the exact document or for exact array. But I just noticed that this document doesn't exactly match the existing array. I'll redo it. Do you have any suggestions or something to add or maybe it should to rename the document?

Comment (seeforschauer) [reply integration/shareddata/composites.go:25]: I would suggest `composite` word.

Comment (ribaraka) [reply integration/shareddata/composites.go:25]: `document-composite`?

Comment (seeforschauer) [reply integration/shareddata/composites.go:25]: yep

**Comment by ribaraka** [comment]
> check dirty and some tests you've added are not ok, can you please fix it?

Could you please rerun tests? My tests should pass successfully.

**Comment by seeforschauer** [comment]
confirm, tests are ok
dirty is not ok.

**Comment by ribaraka** [comment]
Excuse me, but what is dirty?

**Comment by AlekSi** [comment]
https://github.com/FerretDB/FerretDB/runs/6009434112?check_suite_focus=true fails. That steps runs `git diff --exit-code` that exits with non-zero code if there are local changes. And those local changes are present because code was re-formatted on CI.

TL;DR Run `bin/task fmt`

File: integration/query_comparision_test.go [hunk 2]
```diff
@@ -39,22 +39,37 @@ func TestQueryComparisonImplicit(t *testing.T) {
 		filter      bson.D
 		expectedIDs []any
 	}{
+		"Document": {
+			filter:      bson.D{{"value", bson.D{{"foo", int32(42)}, {"42", "foo"}, {"array", bson.A{int32(42), "foo", nil}}}}},
```
Comment (AlekSi) [inline integration/query_comparision_test.go:43]: Let's also add a test for comparision with document with the same keys and values, but in different order. The same in other tests

File: integration/query_comparision_test.go [hunk 3]
```diff
@@ -39,22 +39,37 @@ func TestQueryComparisonImplicit(t *testing.T) {
...
+			expectedIDs: []any{"document-composite"},
+		},
+
+		"Array": {
+			filter:      bson.D{{"value", bson.A{int32(42), "foo", nil}}},
```
Comment (AlekSi) [inline integration/query_comparision_test.go:48]: Same there. The same in other tests

File: integration/query_comparision_test.go [hunk 4]
```diff
@@ -39,22 +39,37 @@ func TestQueryComparisonImplicit(t *testing.T) {
...
+			expectedIDs: []any{"array-three"},
+		},
+
```
Comment (AlekSi) [inline integration/query_comparision_test.go:51]: No empty line there; we group by argument type

File: internal/handlers/common/compare.go [hunk 5]
```diff
@@ -188,15 +189,24 @@ func compare(docValue, filter any) compareResult {
 
 	switch docValue := docValue.(type) {
 	case *types.Document:
+		filter, isFilterDocument := filter.(*types.Document)
```
Comment (AlekSi) [inline internal/handlers/common/compare.go:192]:
```suggestion
		filter, ok := filter.(*types.Document)
```

"comma ok" is a Go idiom

File: internal/handlers/common/compare.go [hunk 6]
```diff
@@ -188,15 +189,24 @@ func compare(docValue, filter any) compareResult {
...
+		if isFilterDocument {
+			return compareDocuments(filter, docValue)
+		}
 		return notEqual
 
 	case *types.Array:
+		switch filter.(type) {
```
Comment (AlekSi) [inline internal/handlers/common/compare.go:199]:
```suggestion
		switch filter := filter.(type) {
```

Then `filter` can be used without type assertion below.

But then why we use different type checking style there and for document above? Please use the same, which is more common in that file

**Review by seeforschauer — APPROVED** [review]
:+1:

File: internal/handlers/common/compare.go [hunk 7]
```diff
@@ -188,15 +189,24 @@ func compare(docValue, filter any) compareResult {
...
+			return compareArray(filter.(*types.Array), docValue)
+		}
+
 		for i := 0; i < docValue.Len(); i++ {
 			arrValue := must.NotFail(docValue.Get(i))
 
 			switch arrValue.(type) {
 			case *types.Document, *types.Array:
-				return notEqual
+				continue
```
Comment (AlekSi) [inline internal/handlers/common/compare.go:209]: That looks very suspicious. `compare` checks if values are equal, not equal, less or greater. It does not checks conditions like "a is a subset of b". I'm not sure why we want to skip some values there.

File: internal/handlers/common/compare.go [hunk 8]
```diff
@@ -253,3 +263,30 @@ func compareOrdered[T constraints.Ordered](a, b T) compareResult {
 func compareNumbers(a float64, b int64) compareResult {
 	return compareOrdered(a, float64(b))
 }
+
+// compareDocuments converts two documents to fjson and compares them.
```
Comment (AlekSi) [inline internal/handlers/common/compare.go:267]: The comparision of two values should compare them directly, not their serialized forms.
Let's add TODO itdem to improve that in the future.

File: internal/handlers/common/compare.go [hunk 9]
```diff
@@ -253,3 +263,30 @@ func compareOrdered[T constraints.Ordered](a, b T) compareResult {
...
+	if string(must.NotFail(fjson.Marshal(a))) == string(must.NotFail(fjson.Marshal(b))) {
+		return equal
+	}
+
+	return notEqual
+}
+
+// compareArray compares two arrays and returns document where array equals exactly the specified array
+// or array contains an element that equals the array.
+func compareArray(filterArr, docArr *types.Array) compareResult {
```
Comment (AlekSi) [inline internal/handlers/common/compare.go:278]: That's a good description but a bad function name.

Let's make `compareArrays` that compares that arrays are exactly the same.

Let's then make `matchArrays` that does the "smart" thing

File: integration/query_comparision_test.go (hunk 2 above)
Comment (ribaraka) [reply integration/query_comparision_test.go:43]: Will do.

File: internal/handlers/common/compare.go (hunk 9 above)
Comment (ribaraka) [reply internal/handlers/common/compare.go:278]: This func actually is going to do the smart thing when it process cases with $gt, $lt, etc. It's just has not been implemented yet. I'm gonna add more logic to this func when I finish with simple $eq and implicit filter. Or I'm adding handler for $gt operator to this PR. WDYT?

Comment (AlekSi) [reply internal/handlers/common/compare.go:278]: I think all `compareXXX` functions should not have any smarts. We almost could call them `equalXXX` except that they also check if a is less or greater than b. We need a separate function when we need to compare that documents and arrays are strictly equal. For example, we might end up deleting our custom functions that do that in tests: https://github.com/FerretDB/FerretDB/blob/main/internal/util/testutil/equal.go

MongoDB documentation calls those smarts "matching", so that could be a good prefix for functions with smarts.

Comment (AlekSi) [reply internal/handlers/common/compare.go:278]: Another options is to move `compare` function for arrays and documents to types.Array and types.Document methods, but that's not possible for scalar types like int64.

Anyway, let's do all that later. Right now, we should separate simple comparision and smart comparision functions

File: internal/handlers/common/compare.go (hunk 7 above)
Comment (AlekSi) [reply internal/handlers/common/compare.go:209]: I guess we should apply the same logic as to array then with `compare` and `match` functions

File: integration/query_comparision_test.go (hunk 2 above)
Comment (ribaraka) [reply integration/query_comparision_test.go:43]: Done

File: integration/query_comparision_test.go (hunk 3 above)
Comment (ribaraka) [reply integration/query_comparision_test.go:52]: Added test with different order of values

File: integration/query_comparision_test.go (hunk 4 above)
Comment (ribaraka) [reply integration/query_comparision_test.go]: fixed

File: internal/handlers/common/compare.go (hunk 5 above)
Comment (ribaraka) [reply internal/handlers/common/compare.go]: Noted. The lines have been removed. I will handle this case in another PR.

File: internal/handlers/common/compare.go (hunk 6 above)
Comment (ribaraka) [reply internal/handlers/common/compare.go]: Noted and fixed

File: internal/handlers/common/compare.go (hunk 8 above)
Comment (ribaraka) [reply internal/handlers/common/compare.go]: I have processed this case, pls have a look.

File: internal/handlers/common/compare.go [hunk 10]
```diff
@@ -253,3 +257,45 @@ func compareOrdered[T constraints.Ordered](a, b T) compareResult {
...
+	}
+	if b == nil {
+		log.Panicf("%v is nil", b)
+	}
+
+	keys := a.Keys()
+	if !slices.Equal(keys, b.Keys()) {
+		return false
+	}
+	return reflect.DeepEqual(a.Map(), b.Map())
```
Comment (seeforschauer) [inline internal/handlers/common/compare.go:274]: no, reflect is less effective, can we use here our `deepEqual`, some code is already written for it.

Comment (ribaraka) [reply internal/handlers/common/compare.go:274]: I can't find our `deepEqual`, could you please guide me?

Comment (seeforschauer) [reply internal/handlers/common/compare.go:274]: it needs to be written

Comment (ribaraka) [reply internal/handlers/common/compare.go:274]: got it. I am going to come up with something :)

Comment (AlekSi) [reply internal/handlers/common/compare.go:274]: It is less about effectiveness and more about correctness. For example, reflect.DeepEqual doesn’t work like we need it to work for our beloved NaNs.

There is a helper in testutil. Please copy and adopt it; I will refactor those copies later

Comment (AlekSi) [reply internal/handlers/common/compare.go:274]: (you can’t use it as-is because testutil depends on testing package, and we don’t want to import in non-test code)

Comment (ribaraka) [reply internal/handlers/common/compare.go:274]: OK! Thanks

**Review by seeforschauer — CHANGES_REQUESTED** [review]
`deepEqual`  instead of reflect

**Review by seeforschauer — APPROVED** [review]
let's merge it and after make a separate PR with `deepEquals`

**Comment by AlekSi** [comment]
#481

**Review by AlekSi — APPROVED** [review]

**Comment by AlekSi** [comment]
I will fix tests in a separate PR ASAP
